File: producer-consumer.go
package main

import "fmt"

var quit = make(chan string)

func consume(buffer chan int) {
	for num := range buffer {
		fmt.Println("Consumed ", num)
		if num == 9999 {
			quit <- "done"
		}
	}
}

func produce(buffer chan int) {
	num := 0
	for num < 10000 {
		buffer <- num
		num = num + 1
	}
}

func main() {
	buffer_size := 10
	ch := make(chan int, buffer_size)
	go produce(ch)
	go consume(ch)
	select {
	case <-quit:
		return
	} // run forever
}

File: producer-consumer.py
from threading import Thread, Semaphore, Lock

buffer = []
buffer_size = 10
filled = Semaphore(0)
empty = Semaphore(buffer_size)
lock = Lock()

class Producer(Thread):
    def run(self):
        num = 0
        while True:
            empty.acquire()
            lock.acquire()
            buffer.append(num)
            lock.release()
            num = num + 1
            filled.release()


class Consumer(Thread):
    def run(self):
        num = 0
        while True:
            filled.acquire()
            lock.acquire()
            num = buffer.pop(0)
            lock.release()
            print("Consumed %i" % num)
            empty.release()


prod = Producer()
cons = Consumer()
prod.start()
cons.start()
    